sdk/lib/pseh/i386/pseh3.c

```diff
@@ -66,10 +68,49 @@ __attribute__((regparm(1)))
 _SEH3$_Unregister(
     volatile SEH3$_REGISTRATION_FRAME *Frame)
 {
-    if (Frame->Handler)
-        _SEH3$_UnregisterFrame(Frame);
+    volatile SEH3$_REGISTRATION_FRAME *CurrentFrame;
+
+    /* Check if the frame has a handler (then it's the registered frame) */
+    if (Frame->Handler != NULL)
+    {
+        /* There shouldn't be any more nested try-level frames */
+        ASSERT(Frame->EndOfChain == Frame);
+
+        /* During unwinding on Windows ExecuteHandler2 installs it's own EH frame,
+           so there can be one or even multiple frames before our own one and we
+           need to search for the link that points to our head frame. */
+        CurrentFrame = (SEH3$_REGISTRATION_FRAME*)__readfsdword(0);
+        if (Frame == CurrentFrame)
+        {
+            /* The target frame is the first installed frame, remove it */
+            __writefsdword(0, (ULONG)Frame->Next);
+            return;
+        }
+
+        /* Loop to find the frame that links the target frame */
+        while (CurrentFrame->Next != Frame)
+        {
+            CurrentFrame = CurrentFrame->Next;
+        }
+
+        /* Remove the frame from the linked list */
+        CurrentFrame->Next = Frame->Next;
+    }
     else
-        _SEH3$_UnregisterTryLevel(Frame);
+    {
+        /* Search for the head frame */
+        CurrentFrame = Frame->Next;
+        while (CurrentFrame->Handler == NULL)
+        {
+            CurrentFrame = CurrentFrame->Next;
+        }
+
+        /* Make sure the head frame points to our frame */
+        ASSERT(CurrentFrame->EndOfChain == Frame);
+
+        /* Remove this frame from the internal linked list */
+        CurrentFrame->EndOfChain = Frame->Next;
```

Comment:
Looking at the [definitions for SEH3 UnregisterFrame and UnregisterTryLevel](https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/lib/pseh/include/pseh/pseh3.h;hb=8fbbb41a19e615b58a17ea8d36461800cc769828#l90), who is now in charge of fiddling with the `%%fs:0` when `_SEH3$_Unregister()` is invoked? Nobody? or (in the case of Windows) the "language specific exception handler function" that you mention in the commit description?

Reply:
ah thanks